### ISSUE TITLE
Provide Git information on runtime (Alternative to SpringBoot build-info)

### ISSUE THREAD

**Issue by DanelRod** [quarkusio/quarkus#29995]
### Description

On a Spring-Boot very easy are able to see information about build and git branch 
![image](https://user-images.githubusercontent.com/115703261/208839485-0050d1a1-4bb0-45d2-803d-4f2fd39258fd.png)

On Quarkus we do not yet have this possibility this is mainly needed when projects have test environments to help for a team know what version and latest build was deployed, especially if Your build may be deployed manually by anyone ( QA/PO )

Spring-Boot build info example: https://docs.spring.io/spring-boot/docs/2.1.13.RELEASE/maven-plugin/examples/build-info.html

### Implementation ideas

Suggest to implement this functionaly under exising quarkus-boom (default's) 
or
May be implemented similar plugin to https://plugins.gradle.org/plugin/com.gorylenko.gradle-git-properties

**Comment by geoand** [comment]
@gastaldi do you remember if we had discussed this in the past or if there is some Quarkiverse extension?

**Comment by gastaldi** [comment]
@geoand nope, I don't remember discussing this and I haven't seen an extension that does that.

**Comment by geoand** [comment]
Thanks for info

**Comment by gsmet** [comment]
Same here. Looks worthwhile though.

**Comment by famod** [comment]
Sounds familiar: #13598

**Comment by justinasdauksa** [comment]
What's the status of this? Would be useful to have.

**Comment by geoand** [comment]
I agree this would be nice to have. Does anyone have any string opinions on what the endpoint(s) should be? Do we want to do something new or just follow Spring Boot's lead (FWIW, Micronaut seems to do something similar)?

Also looping in @radcortez @FroMage @aloubyansky for opinions

**Comment by radcortez** [comment]
I think it makes sense to follow what Spring is already doing. We don't have any restrictions, and it would be familiar to users from the Spring world.

**Comment by aloubyansky** [comment]
The SB's build-info seems to store basic info as properties though, doesn't it?  @DanelRod where is the JSON coming from? I guess I'm missing something.

**Comment by geoand** [comment]
JSON is how SB returns the info IIRC, it's orthogonal to how it's stored

**Comment by aloubyansky** [comment]
Returns from what? Is there an endpoint?

**Comment by geoand** [comment]
Yes, `/info` IIRC

**Comment by aloubyansky** [comment]
Ok, I see. [1] summarizes it. The `build-info` goal does not store any git info, it's just GAV, time, author. For git info there is another (non-DB) plugin. And there is an info endpoint.

[1] https://reflectoring.io/spring-boot-info-endpoint/

**Comment by aloubyansky** [comment]
I am not sure we should be overloading the configs with everything we could. @DanelRod and others, in which way would you like Quarkus to provide this info?

**Comment by geoand** [comment]
I do want to generate the info at build time, but the way I am thinking about it, it won't be totally static, since I want to allow user code to contribute data.

**Comment by geoand** [comment]
Yes exactly, it's extremely similar

**Comment by aloubyansky** [comment]
A CDI bean is an interesting idea.

**Comment by gastaldi** [comment]
Perhaps we can prototype something in a separate extension. I am not 100% convinced this needs to be implemented as a core extension/feature

**Comment by gastaldi** [comment]
I'd vote to close this issue as a duplicate of #13598 and continue the conversation there

**Comment by geoand** [comment]
> I'd vote to close this issue as a duplicate of #13598 and continue the conversation there

I'd say as this one has a lot more info, we close the other one in favor of this (and perhaps rename this one).

**Comment by geoand** [comment]
> Perhaps we can prototype something in a separate extension. I am not 100% convinced this needs to be implemented as a core extension/feature

I can do something in a couple weeks (as next week I'll most be traveling)

**Comment by aloubyansky** [comment]
Keep in mind we already have `quarkus:info` and it's going to evolve as well. The info command is meant to provide Quarkus-specific info. Perhaps we could combine the two at least in terms of data collection and API to represent it.

**Comment by geoand** [comment]
Yeah, that sounds like a very good idea

**Comment by laurentperez** [comment]
if you allow user code to contribute data you'd have to be careful to mask certain contributed values with a placeholder like "*******"

SB uses a sanitizer for /env endpoint see https://github.com/spring-projects/spring-boot/blob/main/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/env/EnvironmentEndpoint.java

Their /info endpoint does not sanitize because exposed info is not provided by user code https://github.com/spring-projects/spring-boot/tree/main/spring-boot-project/spring-boot-actuator/src/main/java/org/springframework/boot/actuate/info

**Comment by geoand** [comment]
@laurentperez do you happen to know the rationale for not allowing user code to contribute in Spring Boot's `/info`?

**Comment by geoand** [comment]
Ah, nice find.

We can do much better I would say 😉

**Comment by laurentperez** [comment]
> @laurentperez do you happen to know the rationale for not allowing user code to contribute in Spring Boot's `/info`?

it is allowed, `InfoContributor` is a public interface and authors can write their own contributors

so it's the author responsability to expose safe values in /info

it is more or less related to https://github.com/quarkusio/quarkus/issues/23429#issuecomment-1148546347 . if management endpoint is exposed on a dedicated port then it's less prone to expose unsafe values

**Comment by DanelRod** [comment]
> I am not sure we should be overloading the configs with everything we could. @DanelRod and others, in which way would you like Quarkus to provide this info?

can be done in the same way as it is provided by spring separate `/info` endpoint. 

mainly this kinda endpoint are used to know what currently is deployed to Your test environment, as all of us working mainly on microservices architecture and with many test environments sometimes is pretty  hard to track what actually deployed where and what was latest build ( especially if You have in Your team technical QA who are able to build any docker image to You'r test machine ), so mainly you should have access to k8s or CI/CD pipelines to know what and where was deployed, but from plain application view - not :)

**Comment by geoand** [comment]
We need to wait for @cescoffier to finish up https://github.com/quarkusio/quarkus/issues/23429 before starting work on this

**Comment by geoand** [comment]
I'll look into this this week. Hopefully we can have something (although perhaps not with every feature mentioned here) for `3.0.0.CR1`

**Comment by geoand** [comment]
I opened https://github.com/quarkusio/quarkus/pull/32161 that proposes a first version that provides the basics of an `/info` endpoint.
If you folks can take a look, it would be great because having this in `3.0.0.CR1` would be nice.

**Comment by mkouba** [comment]
@geoand @gsmet I wonder if we should log a big warning when the app starts, i.e. that `quarkus-info` extension is present and the `q/info` got added and represents a potential security risk.

**Comment by geoand** [comment]
I don't see why we would do that as users are likely used to this from other frameworks

**Comment by mkouba** [comment]
> I don't see why we would do that as users are likely used to this from other frameworks.
> 
> Furthermore, users need to actually add the extension, so they are already opting in.

Well, no matter how used you are to this specific feature if you add this extension and then forget to disable it in production it's a security risk.

**Comment by mkouba** [comment]
> It may or may not be a security risk, but it's opt in, it's not something that's on by default.

Yes, it's opt-in but the usual workflow would be (1) add the extension if needed and (2) disable the extension in production. My point is that you might forget to do (2) and then expose sensitive information (quarkus is used, if your project is OSS then which revision is used, etc.). But it's probably nothing serious so feel free to ignore me :shrug:

**Comment by geoand** [comment]
I don't want to ignore any opinions, I just don't agree in this case :)

**Comment by famod** [comment]
FWIW, I'm +0,5 for such a warning, but only if it can be disabled easily and doesn't nag users in dev mode OOTB.

**Comment by mkouba** [comment]
> FWIW, I'm +0,5 for such a warning, but only if it can be disabled easily and doesn't nag users in dev mode OOTB.

E.g. only log if `LaunchMode.current() == NORMAL`...
